File: internal/application/dto/portfolio.go
package dto

// PortfolioInitRequest 是首次录入本地账户事实的请求。
type PortfolioInitRequest struct {
	Cash        float64         `json:"cash"`
	TotalAssets float64         `json:"total_assets"`
	Positions   []PositionInput `json:"positions"`
}

type PortfolioAdjustmentRequest struct {
	Cash         float64         `json:"cash"`
	TotalAssets  float64         `json:"total_assets"`
	AdjustReason string          `json:"adjust_reason"`
	Positions    []PositionInput `json:"positions"`
}

type PositionInput struct {
	Symbol        string  `json:"symbol"`
	Name          string  `json:"name"`
	Quantity      float64 `json:"quantity"`
	CostPrice     float64 `json:"cost_price"`
	CurrentPrice  float64 `json:"current_price"`
	BuyDate       string  `json:"buy_date,omitempty"`
	PositionState string  `json:"position_state,omitempty"`
	BuyReason     string  `json:"buy_reason,omitempty"`
	AssetTag      string  `json:"asset_tag,omitempty"`
}

type HoldingEditRequest struct {
	PositionID   string        `json:"position_id,omitempty"`
	Reason       string        `json:"reason"`
	Confirmation string        `json:"confirmation"`
	Position     PositionInput `json:"position"`
}

type HoldingRemoveRequest struct {
	PositionID   string `json:"position_id"`
	Reason       string `json:"reason"`
	Confirmation string `json:"confirmation"`
}

type OfflineTransactionRequest struct {
	OperationType string  `json:"operation_type"`
	Symbol        string  `json:"symbol"`
	Name          string  `json:"name,omitempty"`
	Quantity      float64 `json:"quantity"`
	Price         float64 `json:"price"`
	Fees          float64 `json:"fees,omitempty"`
	ExecutedAt    string  `json:"executed_at"`
	Note          string  `json:"note,omitempty"`
	BuyReason     string  `json:"buy_reason,omitempty"`
	AssetTag      string  `json:"asset_tag,omitempty"`
}

type BatchImportValidationRequest struct {
	Rows []BatchImportRow `json:"rows"`
}

type BatchImportRow struct {
	RowNumber     int     `json:"row_number"`
	RowType       string  `json:"row_type"`
	Symbol        string  `json:"symbol"`
	Name          string  `json:"name,omitempty"`
	Quantity      float64 `json:"quantity"`
	CostPrice     float64 `json:"cost_price,omitempty"`
	CurrentPrice  float64 `json:"current_price,omitempty"`
	OperationType string  `json:"operation_type,omitempty"`
	Price         float64 `json:"price,omitempty"`
	Fees          float64 `json:"fees,omitempty"`
	OccurredAt    string  `json:"occurred_at,omitempty"`
	BuyDate       string  `json:"buy_date,omitempty"`
	BuyReason     string  `json:"buy_reason,omitempty"`
	PositionState string  `json:"position_state,omitempty"`
	AssetTag      string  `json:"asset_tag,omitempty"`
}

type BatchImportValidationResponse struct {
	ImportBatchID string                       `json:"import_batch_id"`
	Summary       BatchImportValidationSummary `json:"summary"`
	Rows          []BatchImportRowResult       `json:"rows"`
}

type BatchImportValidationSummary struct {
	RowCount     int `json:"row_count"`
	ValidCount   int `json:"valid_count"`
	InvalidCount int `json:"invalid_count"`
}

type BatchImportRowResult struct {
	RowNumber int      `json:"row_number"`
	Valid     bool     `json:"valid"`
	Errors    []string `json:"errors,omitempty"`
}

type BatchImportConfirmRequest struct {
	ImportBatchID string           `json:"import_batch_id"`
	Rows          []BatchImportRow `json:"rows"`
	ConfirmReason string           `json:"confirm_reason"`
}

type CorrectionRequest struct {
	TargetType       string `json:"target_type"`
	TargetID         string `json:"target_id"`
	BeforeJSON       string `json:"before_json"`
	AfterJSON        string `json:"after_json"`
	CorrectionReason string `json:"correction_reason"`
}

type RebalanceReviewRequest struct {
	TargetCoreRatio      float64 `json:"target_core_ratio"`
	TargetSatelliteRatio float64 `json:"target_satellite_ratio"`
	TargetCashRatio      float64 `json:"target_cash_ratio"`
	DriftThreshold       float64 `json:"drift_threshold,omitempty"`
	ReviewDate           string  `json:"review_date,omitempty"`
}

type RebalanceReviewResponse struct {
	ReviewID        string                `json:"review_id"`
	ReviewDate      string                `json:"review_date"`
	TotalAssets     float64               `json:"total_assets"`
	DriftThreshold  float64               `json:"drift_threshold"`
	Items           []RebalanceReviewItem `json:"items"`
	AuditEventIDs   []string              `json:"audit_event_ids"`
	SafetyStatement string                `json:"safety_statement"`
}

type RebalanceReviewItem struct {
	Bucket         string  `json:"bucket"`
	TargetRatio    float64 `json:"target_ratio"`
	ActualRatio    float64 `json:"actual_ratio"`
	DriftRatio     float64 `json:"drift_ratio"`
	TargetValue    float64 `json:"target_value"`
	ActualValue    float64 `json:"actual_value"`
	Recommendation string  `json:"recommendation"`
	ManualAmount   float64 `json:"manual_amount"`
}

type LocalFactWriteResponse struct {
	SnapshotID      string   `json:"snapshot_id,omitempty"`
	PositionID      string   `json:"position_id,omitempty"`
	TransactionID   string   `json:"transaction_id,omitempty"`
	ImportBatchID   string   `json:"import_batch_id,omitempty"`
	CorrectionID    string   `json:"correction_id,omitempty"`
	AuditEventIDs   []string `json:"audit_event_ids"`
	SafetyStatement string   `json:"safety_statement"`
}

type PortfolioWriteResponse struct {
	SnapshotID            string   `json:"snapshot_id"`
	PositionCount         int      `json:"position_count"`
	PositionSnapshotCount int      `json:"position_snapshot_count,omitempty"`
	AuditEventIDs         []string `json:"audit_event_ids"`
}

type PortfolioCurrentResponse struct {
	Snapshot  PortfolioSnapshotDTO `json:"snapshot"`
	Positions []PositionDTO        `json:"positions"`
}

type PortfolioSnapshotDTO struct {
	SnapshotID    string  `json:"snapshot_id"`
	SnapshotTime  string  `json:"snapshot_time"`
	Cash          float64 `json:"cash"`
	TotalAssets   float64 `json:"total_assets"`
	CashRatio     float64 `json:"cash_ratio"`
	HighRiskRatio float64 `json:"high_risk_ratio"`
	PositionCount int     `json:"position_count"`
}

type PositionDTO struct {
	PositionID            string  `json:"position_id"`
	Symbol                string  `json:"symbol"`
	Name                  string  `json:"name"`
	Quantity              float64 `json:"quantity"`
	CostPrice             float64 `json:"cost_price"`
	CurrentPrice          float64 `json:"current_price"`
	MarketValue           float64 `json:"market_value"`
	UnrealizedProfitRatio float64 `json:"unrealized_profit_ratio"`
	PositionState         string  `json:"position_state"`
	BuyDate               string  `json:"buy_date,omitempty"`
	BuyReason             string  `json:"buy_reason,omitempty"`
	AssetTag              string  `json:"asset_tag,omitempty"`
}

File: web/src/types/portfolio.ts
export interface PortfolioCurrentResponse {
  snapshot: PortfolioSnapshot
  positions: Position[]
}

export interface PortfolioSnapshot {
  snapshot_id: string
  snapshot_time: string
  cash: number
  total_assets: number
  cash_ratio: number
  high_risk_ratio: number
  position_count: number
}

export interface Position {
  position_id: string
  symbol: string
  name: string
  quantity: number
  cost_price: number
  current_price: number
  market_value: number
  unrealized_profit_ratio: number
  position_state: 'normal' | 'sell_only' | 'frozen_watch' | string
  buy_date?: string
  buy_reason?: string
  asset_tag?: string
}

export interface PositionInput {
  symbol: string
  name: string
  quantity: number
  cost_price: number
  current_price: number
  buy_date?: string
  position_state?: 'normal' | 'sell_only' | 'frozen_watch' | string
  buy_reason?: string
  asset_tag?: string
}

export interface PortfolioInitRequest {
  cash: number
  total_assets: number
  positions: PositionInput[]
}

export interface PortfolioAdjustmentRequest extends PortfolioInitRequest {
  adjust_reason: string
}

export interface PortfolioWriteResponse {
  snapshot_id: string
  position_count: number
  position_snapshot_count?: number
  audit_event_ids: string[]
}

export interface LocalFactWriteResponse {
  snapshot_id?: string
  position_id?: string
  transaction_id?: string
  import_batch_id?: string
  correction_id?: string
  audit_event_ids: string[]
  safety_statement: string
}

export interface HoldingEditRequest {
  position_id?: string
  reason: string
  confirmation: string
  position: PositionInput
}

export interface HoldingRemoveRequest {
  position_id: string
  reason: string
  confirmation: string
}

export interface OfflineTransactionRequest {
  operation_type: 'buy' | 'sell' | 'reduce' | string
  symbol: string
  name?: string
  quantity: number
  price: number
  fees?: number
  executed_at: string
  note?: string
  buy_reason?: string
  asset_tag?: string
}

export interface BatchImportRow {
  row_number?: number
  row_type: 'holding' | 'transaction' | string
  symbol: string
  name?: string
  quantity: number
  cost_price?: number
  current_price?: number
  operation_type?: 'buy' | 'sell' | 'reduce' | string
  price?: number
  fees?: number
  occurred_at?: string
  buy_date?: string
  buy_reason?: string
  position_state?: 'normal' | 'sell_only' | 'frozen_watch' | string
  asset_tag?: string
}

export interface BatchImportValidationRequest {
  rows: BatchImportRow[]
}

export interface BatchImportValidationSummary {
  row_count: number
  valid_count: number
  invalid_count: number
}

export interface BatchImportRowResult {
  row_number: number
  valid: boolean
  errors?: string[]
}

export interface BatchImportValidationResponse {
  import_batch_id: string
  summary: BatchImportValidationSummary
  rows: BatchImportRowResult[]
}

export interface BatchImportConfirmRequest extends BatchImportValidationRequest {
  import_batch_id: string
  confirm_reason: string
}

export interface CorrectionRequest {
  target_type: 'portfolio_snapshot' | 'position' | 'position_snapshot' | 'position_transaction' | 'import_batch' | string
  target_id: string
  before_json: string
  after_json: string
  correction_reason: string
}

export interface RebalanceReviewRequest {
  target_core_ratio: number
  target_satellite_ratio: number
  target_cash_ratio: number
  drift_threshold?: number
  review_date?: string
}

export interface RebalanceReviewItem {
  bucket: 'core' | 'satellite' | 'cash' | string
  target_ratio: number
  actual_ratio: number
  drift_ratio: number
  target_value: number
  actual_value: number
  recommendation: string
  manual_amount: number
}

export interface RebalanceReviewResponse {
  review_id: string
  review_date: string
  total_assets: number
  drift_threshold: number
  items: RebalanceReviewItem[]
  audit_event_ids: string[]
  safety_statement: string
}
